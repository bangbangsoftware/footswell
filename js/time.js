export function timer(minutesElement, secondsElement, underLimit = 10) {
  const increment = incrementFn(minutesElement, secondsElement, underLimit);
  const reset = resetFn(minutesElement, secondsElement, underLimit);
  return {
    increment,
    reset
  };
}

const incrementFn = (minutesElement, secondsElement, underLimit) => () => {
  const secs = parseInt(secondsElement.innerText) + 1;
  if (secs > 59) {
    incrementSeconds(minutesElement, secondsElement, underLimit);
    return;
  }
  secondsElement.innerText = secs < 10 ? `0${secs}` : secs;
};

const resetFn = (minutesElement, secondsElement) => () => {
  secondsElement.innerText = "00";
  secondsElement.classList.remove("red");
  minutesElement.innerText = "00";
  minutesElement.classList.remove("red");
};

export function timeFormat(date = new Date()) {
  const hr = zeroFill(date.getHours());
  const mn = zeroFill(date.getMinutes());
  const sc = zeroFill(date.getSeconds());
  return `${hr}:${mn}:${sc}`;
}

const incrementSeconds = (minutes, seconds, underLimit) => {
  seconds.innerText = 0;
  const mins = parseInt(minutes.innerText) + 1;
  if (mins < underLimit) {
    minutes.innerText = "0" + mins;
    return;
  }
  minutes.innerText = zeroFill(mins);
  seconds.classList.add("red");
  minutes.classList.add("red");
};

const zeroFill = i => {
  if (i < 10) {
    return "0" + i;
  }
  return i;
};
