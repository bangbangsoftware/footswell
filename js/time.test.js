import { timer, timeFormat } from "./time";

describe("Timer...", () => {
  const classReg = {};
  const classList = id => {
    return { remove: () => (classReg[id] = {}), add: t => (classReg[id] = t) };
  };
  const minutesElement = { innerText: 0, classList: classList("m") };
  const secondsElement = { innerText: 0, classList: classList("s") };
  const underLimit = 2;
  const clock = timer(minutesElement, secondsElement, underLimit);

  it("should get a clock", () => {
    expect(clock).toBeDefined();
  });

  // increment removed and replaced by adjust which works relative to a date
  xit("should be able to increment", () => {
    clock.increment();
    expect(secondsElement.innerText).toBe("01");
  });

  xit("should be able to increment to a minute", () => {
    for (let x = 0; x < 60; x++) {
      clock.increment();
    }
    expect(secondsElement.innerText).toBe("01");
    expect(minutesElement.innerText).toBe("01");

    for (let x = 0; x < 60; x++) {
      clock.increment();
    }
    expect(secondsElement.innerText).toBe("01");
    expect(minutesElement.innerText).toBe("02");
    expect(classReg.m).toBe("red");
    expect(classReg.s).toBe("red");

    clock.reset();
    expect(secondsElement.innerText).toBe("00");
    expect(minutesElement.innerText).toBe("00");
    expect(classReg.s+"").toBe(""+{});
  });

  it("should be able to render some time", () => {
    const toRender = new Date("July 21, 2020 08:00:00");
    expect(timeFormat(toRender)).toMatchSnapshot();
  });
});
