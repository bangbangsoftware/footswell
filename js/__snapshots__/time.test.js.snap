// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer... should be able to render some time 1`] = `"08:00:00"`;
